Generate RPC Services from Go interfaces

dennwc opened the pull request:
Generate RPC services from Go interfaces. Also make transform to generate service definition instead of a list of RPCs.

dennwc committed:
generate rpc services from interfaces and adapters to Go interfaces

dennwc commented:
There will be some more changes to generate client code as well, so don't merge yet.

Serabe commented:
We've been talking about this PR IRL. The main problem with this is that interfaces are the translations of `oneof` in protobuf. If in a future we want to support them somehow, we would have closed that door by accepting this.

We like the idea, though we would need to find a workaround for this. We are considering options for the comment `proteus:generate` but haven't settled in anything yet (ideas more than welcomed here).

Also, there is the problem that imagine you have the following interface:

```go
type Person interface {
  Name() string
}
```

This would generate a `personRemote` struct satisfying this interface that would swallow any error in the called.

My initial proposal would be to apply the following two changes:
- For an interface to be generated as a service, it needs to be called with `//proteus:generate:service`. This would imply no changes in `hasGenerateComment` since it checks with `strings.HasPrefix`.
- For an interface to be generated as a service, all the methods need to return, at least, an error in final position. That would allow us to work with a remote version being able to handle the errors.

As before, an example needs to be added in `example/models.go` and checked in `example_test.go`

What do you think, @dennwc, @erizocosmico, and @mcuadros?

Serabe commented on protobuf/transform.go:
```diff
@@ -128,6 +144,21 @@ func (t *Transformer) transformFunc(pkg *Package, f *scanner.Func, names nameSet
+		}
+	}
+	if len(srv.RPCs) == 0 {
+		return nil
```
There is no test checking what happens if we generate an empty interface. Can we add one to make sure nothing explodes?

erizocosmico commented:
@Serabe why forcing that all interface methods return error? Currently we don't enforce that with functions or methods. So, why do it for interfaces?

dennwc commented:
@Serabe I agree with @erizocosmico that logic should match current behavior for global functions. If user wants an interface function with no error, he should be aware that RPC can fail and zero result will be returned. This can be documented, but is a valid use case.

Regarding type unions (interfaces that are implemented by multiple message types), I think we can easily distinguish between them and service interfaces. In any case RPC generation from interfaces requires explicit `//proteus:generate` annotation. And type unions (generated for protobuf), are useless without a message that contains them as a field type. And if we find a message/struct with such interface we can generate `oneof` automatically, the same way as dependent structs are generated.

And sure, I will add examples, but I need to commit few more changes first to integrate context support.

---

Review status: 0 of 13 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/src-d/proteus/89#-:-KoIUbYKB_xrFohnmsAM:b7ugdru)*
<!-- Sent from Reviewable.io -->

Serabe commented:
The matching behaviour for [current functions is with an error as last returned value in the client](https://github.com/src-d/proteus/blob/a40ce5fe0e625d6c1e0abe20fcb59f6653e09b42/example/generated.pb.go#L231-L239). We allow functions without errors as last returned value because we only generate the server, and therefore, we don't force functions to return an error.

```go
func A(b string) int {
    return 0
}

func AWithError(b string) (int, error) {
    return A(b), nil
}
```

Since we are going from not having an error (original function) to having an error (in order to satisfy the signature in the server's interface) we are no loosing any information.

This PR adds a sort of client, the remote struct. In this case we have the following:

```go
//proteus:generate
type RemoteType interface {
    Name() string
}
```

And we are generating something like:

```go
func (c *remoteTypeRemote) Name() string {
    s, _ := c.Client.Name(ctx) // here we discard the error
    return s
}
```

So we are going the opposite direction, from having an error to not having it. That causes some information loss I would rather not have.

Since this is new behaviour, I would like to be conservative so far.

Thank you!

dennwc commented:
I totally understand what you are saying, and I agree that dropping an error in RPC client is a bad idea, but it's not possible to add an error to client's interface definition.

The point I was trying to make is that if client asks to generate an RPC client-server pair for his interface and this interface has a method with no error, RPC client should match this, or it will not be able to implement user's interface. It will render generated code useless in context of user's interface.

I propose printing a warning message in proteus log for these situations and/or generating a FIXME comment in client definition, so user will be aware that omitting an error might cause unexpected behavior.

---

Review status: 0 of 13 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/src-d/proteus/89#-:-KoMSwRg-wS3qdRMblHX:bs3besq)*
<!-- Sent from Reviewable.io -->

dennwc committed:
add examples; fix custom imports resolution

dennwc commented:
Rebased, integrated with context and added examples. @Serabe please take a look.

---

Review status: 0 of 18 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[protobuf/transform.go, line 157 at r2](https://reviewable.io:443/reviews/src-d/proteus/89#-KoIUZVGj3qC1MMCiIdu-r2-157:-KogS3yxfRUG1E6PrOgT:b2w5qpu) ([raw file](https://github.com/src-d/proteus/blob/312025b1805590cb1e7b89b91068af20437640cf/protobuf/transform.go#L157)):*
<details><summary><i>Previously, Serabe (Sergio Arbeo) wrote…</i></summary><blockquote>

There is no test checking what happens if we generate an empty interface. Can we add one to make sure nothing explodes?
</blockquote></details>

Done. Added as an example.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/src-d/proteus/89#-:-KogS6vkxdGdJCJ9nRPa:b-k08ekg)*
<!-- Sent from Reviewable.io -->

dennwc commented:
@Serabe Any updates on this?

Serabe commented:
Still would like to only accept interfaces with the error, unless another solution is proposed. Don't like the idea of losing errors.

dennwc commented:
@Serabe I already proposed two solutions. What about them?

dennwc commented:
Okay, I think I should provide an example to show that this might be a desired behavior for some use cases.

Let's assume we are building a plugin system with microservices. Each plugin can do some useful work, and has an additional method to expose meta-information about itself:

```protobuf
message Info {
  enum State {
     UNAVAILABLE = 0;
     READY = 1;
  }
  State state = 1;
  string name = 2;
}

service Module {
    rpc Info(Empty) return (Info);
    rpc Do(DoReq) returns (DoResp);
} 
```

This corresponds to a following Go interface:
```golang
type Module interface{
   Info() Info
   Do(DoReq) (DoResp, error)
}
```

As you can see, `Info` method has no error. But since default value of `Info` struct already has a valid meaning (status = unavailable), user can still verify that an error happened on the RPC and module cannot be reached.

Serabe commented:
We cannot check if an struct has anything that could be considered an unavailable state like in the `Info` above and still we are losing information (the error returned by gRPC).

To resume: For this PR to be accepted it needs to:

1. Only generate services from interfaces with `//proteus:generate:service` or something that is not a plain `//proteus:generate`.
2. Only generate methods that return an error in last place.

Thank you!
